src/rust/ide/lib/view/src/graph_editor/component/edge.rs

```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
-                let rot    = right - &angle/2.0 + start_angle;
-                let mask   = Plane().cut_angle_fast(angle).rotate(rot);
-                let shape  = ring * mask;
+            ( radius:f32
+            , angle:f32
+            , start_angle:f32
+            , pos:Vector2<f32>
+            , dim:Vector2<f32>
+            , focus_split_center:Vector2<f32>
+            , focus_split_angle:f32
+            ) {
+                let width  = &LINE_WIDTH.px();
+                let shape  = corner_base_shape(&radius,width,&angle,&start_angle);
 
                 let shadow_size = 10.px() + 1.px();
-                let n_radius = &shadow_size + 1.px() * dim.y();
-                let n_shape  = Rect((&shadow_size*2.0 + 2.px() * dim.x(),&n_radius*2.0)).corners_radius(n_radius);
-                let n_shape  = n_shape.fill(color::Rgba::new(1.0,0.0,0.0,1.0));
+                let node_radius = &shadow_size + 1.px() * dim.y();
+                let node_shape  = Rect(
+                    (&shadow_size*2.0 + 2.px() * dim.x(),&node_radius*2.0)).corners_radius(node_radius);
+                let node_shape  = node_shape.fill(color::Rgba::new(1.0,0.0,0.0,1.0));
                 let tx       = - 1.px() * pos.x();
                 let ty       = - 1.px() * pos.y();
-                let n_shape  = n_shape.translate((tx,ty));
+                let node_shape  = node_shape.translate((tx,ty));
 
-                let shape = shape * n_shape;
-                let shape = shape.fill(color::Rgba::from($($color)*));
+                let shape = shape.intersection(node_shape);
 
-                shape.into()
+                let split_shape = FocusedEdge::new(
+                shape,&focus_split_center.px(),&focus_split_angle.into());
+                let shape       = split_shape.fill($color, $highlight_color);
+
+                let hover_width = width + HOVER_EXTENSION.px() * 2.0;
+                let hover_area  = corner_base_shape(&radius,&hover_width,&angle,&start_angle);
+                let hover_area  = hover_area.fill(HOVER_COLOR);
+                (hover_area + shape).into()
+            }
+        }
+
+        impl EdgeShape for component::ShapeView<Shape> {
+            fn set_focus_split_center_local(&self, center:Vector2<f32>) {
+                self.shape.focus_split_center.set(center);
+            }
+
+            fn set_focus_split_angle(&self, angle:f32) {
+                 self.shape.focus_split_angle.set(angle);
+            }
+
+            fn sprite(&self) -> &Sprite{
+                &self.shape.sprite
+            }
+
+            fn events(&self) -> &ShapeViewEvents{
+                &self.events
+            }
+
+            fn normal_local(&self, point:Vector2<f32>) -> nalgebra::Rotation2<f32> {
+                point_rotation(point)
+            }
+
+            fn snap_local(&self, point:Vector2<f32>) -> Option<Vector2<f32>> {
+                let radius = self.shape.radius.get();
```

Comment:
why this does not need to have allow(depraecated)?

Reply:
Because we always were able to get the values from shapes. Just not sprites. Should this be different?

Reply:
Interesting. I don't know. I will think about it. Thanks for pointing it out!